docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,238 @@
+# KDS Redesign
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4354
+
+## Context and Problem Statement
+
+The current implementation of KDS is quite complex, and there may be some bugs, such as the one found at https://github.com/kumahq/kuma/pull/5373, which was very difficult to identify and resolve. The use of a bi-directional stream for configuration exchange between the zone and the global can also be challenging to implement. Additionally, KDS operates as a State of The World, meaning that whenever an item in the resource type changes, the entire state is sent. While this doesn't appear to be a problem when synchronizing data from Global to Zone, it may be less efficient when there are many Dataplane changes. There are several improvements we want to make in the future such as:
+* working with different older versions of the control plane on the zone,
+* creating resources on the zone and syncing them to the global,
+* creating separate gRPC services for synchronization between the zone and global,
+* supporting multiple control plane tenants,
+* and sending only changes instead of the state of the world when synchronizing data from the zone to the global.
+
+## Considered Options
+
+* Redesign KDS
+
+## Decision Outcome
+
+Chosen option: "Redesign KDS".
+
+## Solution
+
+Currently, we are using one bidirection stream for sending resources from Zone to Global and Global to Zone.
```

Comment:
Should we plan to do some simple perf measurements to make sure we're indeed improving KDS performance? If so, I think it makes sense to reflect in MADR what we want to measure, i.e:
* Global CP mem/cpu utilization
* Latency

Reply:
I think that would be nice to make some simple perf tests that require to create 1000+ policies and 1000+ dataplanes:
* measure mem/cpu usage when creating 
* we could try tu measure bandwidth
* Latency

Reply:
While perf measurements could be useful (And quite easily feasible with a benchmark test in go -- we likely don't need entire CPs to do this) I think this point is mostly trying to make things less complex not so much increasing perf.